### ISSUE TITLE
[Question] Audio loop using dynamic resources

### ISSUE THREAD

**Issue by vbarthel-fr** [google/oboe#176]
Hello!

I have a general question regarding the use of "dynamic" components inside an audio loop. Please do not hesitate to ask me if this is the wrong place to get this kind of information :)

What are the possible options to modify resources that are used by the audio thread from another thread?

In other words, what are the solutions to dynamically manage resources that are shared by the audio thread and another thread? For example, if I have some audio components that use kind a lot of RAM/CPU, I could be interested in using the minimum number of those components needed at a given time. That would require the number of resources to be changed at runtime.

Here is a small example to give more context:

Imagine an audio engine with two components - a sampler and a synth - with the following audio loop:

```cpp
DataCallbackResult
AudioEngine::onAudioReady(AudioStream *oboeStream, void *audioData, int32_t numFrames) {
    sampler->renderAudio(audioData, numFrames);
    synth->renderAudio(audioData, numFrames);
    return DataCallbackResult::Continue;
}
```

Now imagine that this synth have a dynamic number of oscillators that can be changed at runtime.

```cpp
void Synth::setNumberOfOscillators(int numberOfOscillators) {
    // pseudo code
    // if numberOfOscillators > currentNumberOfOscillators: allocate more oscillators
    // if numberOfOscillators < currentNumberOfOscillators: free unused oscillators
}

void Synth::renderAudio(void *audioData, int32_t numFrames) {
    // pseudo code
    // for each oscillator: render audio
}
```

`Synth::setNumberOfOscillators` and `Synth::renderAudio` cannot be executed at the same time: they both touch the underlying oscillators. For example, freeing some oscillators while looping though them will lead to a crash.

Unfortunately, the `Synth::renderAudio` is called from the audio thread and the `Synth::setNumberOfOscillators` is called from the main thread (for the sake of this example, let's say that the user can change the number of oscillators from the UI). Eventually, `Synth::setNumberOfOscillators` will be called while the audio thread is executing `Synth::renderAudio` and the app will crash.

## Option 1: locking the access to the oscillators
The access to the oscillators could be guarded by a lock so that two threads can not manipulate them at the same time. But from what I have been told by people more experienced with real time audio application than me, using locks inside the audio loop would be a bad idea and is considered as a bad practice.

## Option 2: pausing the audio loop
Another solution could be to pause the audio loop, apply the changes, and then resume the audio loop:
  
```cpp
void Synth::setNumberOfOscillators(int numberOfOscillators) {
    // pseudo code
    // pause (and actually wait for the audio loop to be paused)
    // https://github.com/google/oboe/blob/6fc2f33094434706de2762eee2644d271d8f5ffe/include/oboe/AudioStream.h#L63
    // if numberOfOscillators > currentNumberOfOscillators: allocate more oscillators
    // if numberOfOscillators < currentNumberOfOscillators: free unused oscillators
    // request start
}

void Synth::renderAudio(void *audioData, int32_t numFrames) {
    // pseudo code
    // for each oscillator: render audio
}
```

Assuming that when `Result AudioStream::pause(int64_t timeout)` returns, the audio thread is actually no longer executing our `onAudioReady` callback, the audio thread would not be touching the oscillators during their allocations/deletions.

This option is not really satisfying: the pause will probably be audible by the user and the sampler will also be affected by this "pause".

## Option 3: deferring the changes to the audio thread

```cpp
void Synth::setNumberOfOscillators(int numberOfOscillators) {
    // pseudo code
    // if targetedNumberOfOscillators == numberOfOscillators: return
    // targetedNumberOfOscillators = numberOfOscillators
}

void Synth::renderAudio(void *audioData, int32_t numFrames) {
    // pseudo code
    // if targetedNumberOfOscillators > currentNumberOfOscillators: allocate more oscillators
    // if targetedNumberOfOscillators < currentNumberOfOscillators: free unused oscillators

    // for each oscillator: render audio
}
```

`Synth::setNumberOfOscillators` no longer applies the new number of oscillators, but instead it writes the new targeted number of oscillators. It's now the audio loop that will be responsible of applying the new number of oscillators. The value of targetedNumberOfOscillators might still be inconsistent between the two threads when the main thread write a new value, but it would not be an issue since the audio loop would apply the changes at the next call.

Since the new number of oscillators does not need to be applied "immediately" but can wait for the next audio render call, having the main thread writing to targetedNumberOfOscillators and the audio thread reading it looks like a good alternative to me.

But I have been told that having memory allocations / deallocations inside the audio loop might not be the best idea. In the case that `Synth::setNumberOfOscillators` is not called too often, this is probably the solution I like the most.

I would be more than happy to have your thoughts on this issue! How would you handle such mechanism? Do you have any alternative that you would recommend for this kind of problem?

Thanks a lot for having read my question until the end :) Please feel free to ask me any additional explanation!

**Comment by philburk** [comment]
>  I have been told that having memory allocations / deallocations inside the audio loop might
> not be the best idea. 

That is an understatement. Allocating or freeing memory from a real-time audio callback is a terrible idea. It will often cause a glitch. Also blocking on a mutex or doing I/O can glitch. The only things you should do in a callback is math and using atomics.

You should allocate all the resource you think you will need BEFORE starting the audio stream.  Then communicate with your audio stream using an atomic FIFO or through atomic variables.  You could send MIDI commands through a FIFO and then parse them in the audio callback. That is usually very quick and provide a clean and narrow interface.

**Comment by vbarthel-fr** [comment]
Thanks a lot for your answer :)

If I understand correctly a better option to the problem illustrated in my first post would be to find a reasonable maximum number of `Synths`, instantiate all of them before actually starting the audio stream and then only loop through the "activated" ones. 

## Option 5: limiting the maximum number of oscillators supported and using atomic variables to communicate with the audio stream 

Inspiration: https://github.com/google/oboe/blob/master/src/fifo/FifoController.h

```cpp
void Synth::setNumberOfOscillators(int32_t numberOfOscillators) {
    // assuming that numberOfOscillators <= maximum number of oscillators
    this.atomicActiveOscillators.store(numberOfOscillators, std::memory_order_release);
}

void Synth::renderAudio(void *audioData, int32_t numFrames) {
    uint32_t numberOfOscillators = this.atomicActiveOscillators.load(std::memory_order_acquire);
    for(uint32_t i = 0; i < numberOfOscillators; i++) {
        this.oscillators[i].renderAudio(audioData, numFrames);
    }
}
```

This solves the high CPU usage problem: only the requested number of Synths will actually render data at a given time. For the high RAM usage, the "problem" remains since no Synths are deallocated when no longer needed. The RAM usage can not be reduced at runtime by setting a smaller number of Synths. At least, while the audio stream is started. But I guess it's much more easier to reduce the RAM usage of the `Synth` class itself with a better design (sharing buffers and other resources with a high RAM usage, flyweight pattern?) than trying to deallocate them in the context of a real time audio application!

Thanks again for your help!

**Comment by philburk** [comment]
>  find a reasonable maximum number of Synths, instantiate all of them before actually
> starting the audio stream and then only loop through the "activated" ones.

Yes, that is the right idea.

If you just want to turn on and off oscillators then that will work OK.  But in a normal synthesizer you would allocate oscillators or voices using a more complex scheme.

Also when you start or stop the oscillators you will get a click unless you ramp them up or down. That requires an envelope.

Also if you are letting each oscillator mix its own output then you must zero the output buffer before calling the oscillator/voice loop.

RAM is not usually a problem for oscillators. There is very little unique state: phase, phaseIncrement, and state related to timbre.

Are you using wavetables or a different technique? Try to share read-only tables as static data.

**Comment by philburk** [comment]
I think this question has now been answered. I hate to close it because the original was so beautifully formatted! But I will close it.
Please open a new issue if you have a new question.
